### PR TITLE
TL/UCP: endpoints hash

### PR DESCRIPTION
**Pull request by vspetrov**
## What
Changes the way ucp ep are stored in TL. Now we use a hash table located on ucc_tl_ucp_contet_t. The hash implementation is a "khash" - opensource free license MIT implementation also used by UCX. 

## Why ?
Storing eps on team results in huge memory leak (#108) and team create overhead doe to connect-disconnect

## How ?
The key to the hash table is ucc_context_id_t structure which consists of a triplet: host_id, pid, and ucc context sequence number.
TL/UCP exchanges ucp_worker addresses + the keys during team creation (exchange during ctx create or PMIx-like iface is not yet implemented). The ep is then created only once and put into hash based on the key.

OVerheads:
1. i've measured the cost of the hash itself (implemented a simple benchmark):
    1.1 for the size of the hash ~1000 the overhead to find the ep in the has is ~0.02 us (<2% compared to network p2p cost)
     1.2 for the size of the hash ~10000 it is ~0.03 us.
     1.3 for the size of the hash ~1000000 it is ~0.08 us.
2. Eval on the Collectives (see xls attached 
[ep_hash_eval.xlsx](https://github.com/openucx/ucc/files/6171030/ep_hash_eval.xlsx). Shortly, for the allreduce, the difference ~5% is observed on 1 node, ~3% on 2 nodes, and nearly 0 difference on 8 nodes. And of course it can only be observed on small messages.

All this makes the approach feasible imho. THe overheads are minimal. Moreover, we can optimize smallest colls (barrier, and small msg ar) by pre-allocating the alg data structure with EPs (doing hash access once during team create) - this will be made in another PR.

### CODE REVIEW

**Review by manjugv — APPROVED** [review]

File: src/components/tl/ucp/tl_ucp_context.c [hunk 1]
```diff
@@ -126,7 +129,12 @@ UCC_CLASS_INIT_FUNC(ucc_tl_ucp_context_t,
 
 UCC_CLASS_CLEANUP_FUNC(ucc_tl_ucp_context_t)
 {
+    ucc_status_t status;
     tl_info(self->super.super.lib, "finalizing tl context: %p", self);
+    while (UCC_OK != (status = ucc_tl_ucp_close_eps(self))) {
+        ; //TODO can we hang the runtime this way ?
+    }
+	kh_destroy(tl_ucp_ep_hash, self->ep_hash);
```
Comment (Sergei-Lebedev) [inline src/components/tl/ucp/tl_ucp_context.c:137]: alignment

File: src/components/tl/ucp/tl_ucp_ep.c [hunk 2]
```diff
@@ -29,47 +30,54 @@ static inline ucc_status_t ucc_tl_ucp_connect_ep(ucc_tl_ucp_context_t *ctx,
...
+        ep = tl_ucp_hash_pop(ctx->ep_hash);
+        if (NULL == ep) {
+            break;
         }
         state->close_req =
-            ucp_ep_close_nb(eps[state->ep], UCP_EP_CLOSE_MODE_FLUSH);
+            ucp_ep_close_nb(ep, UCP_EP_CLOSE_MODE_FLUSH);
         if (UCS_PTR_IS_ERR(state->close_req)) {
             tl_error(ctx->super.super.lib, "failed to start ep close, ep %p",
-                     eps[state->ep]);
```
Comment (Sergei-Lebedev) [inline src/components/tl/ucp/tl_ucp_ep.c:72]: looks like ep field is not needed in ucc_tl_ucp_ep_close_state_t  anymore, should we delete it?

File: src/components/tl/ucp/tl_ucp_ep.c [hunk 3]
```diff
@@ -29,47 +30,54 @@ static inline ucc_status_t ucc_tl_ucp_connect_ep(ucc_tl_ucp_context_t *ctx,
...
             return UCC_INPROGRESS;
         }
         ucp_request_free(state->close_req);
-        eps[state->ep++] = NULL;
     }
-
-    for (; state->ep < n_eps; state->ep++) {
-        if (!eps[state->ep]) {
-            continue;
+    while (1) {
```
Comment (Sergei-Lebedev) [inline src/components/tl/ucp/tl_ucp_ep.c:71]: minor, I would better write it without break, something like
```C
ep = tl_ucp_hash_pop(ctx->ep_hash);
while(ep) {
...
  ep = tl_ucp_hash_pop(ctx->ep_hash);
}
```

File: src/components/tl/ucp/tl_ucp.h [hunk 4]
```diff
@@ -69,15 +70,10 @@ typedef struct ucc_tl_ucp_task ucc_tl_ucp_task_t;
...
-    int                        context_ep_storage; /*< The flag
-              indicates whether ucp endpoints are stored on the
-              ucc_tl_ucp_context or are they created per-team.
-              This optimization is only possible when user provides
-              the necessary rank mappings team_rank->context_rank. */
-    ucp_ep_h                  *eps;
     ucc_rank_t                 size;
     ucc_rank_t                 rank;
     ucc_tl_ucp_addr_storage_t *addr_storage;
+    int                        ctx_addr_exchange;
```
Comment (Sergei-Lebedev) [inline src/components/tl/ucp/tl_ucp.h:76]: what is the meaning of ctx_addr_exchange? is it used?

File: src/components/tl/ucp/tl_ucp_ep.h [hunk 5]
```diff
@@ -10,34 +10,45 @@
...
 ucc_status_t ucc_tl_ucp_connect_ctx_ep(ucc_tl_ucp_context_t *ctx, ucc_rank_t ctx_rank);
 
-ucc_status_t ucc_tl_ucp_close_eps(ucc_tl_ucp_context_t *ctx, ucp_ep_h *eps,
-                                  ucc_rank_t n_eps);
+ucc_status_t ucc_tl_ucp_close_eps(ucc_tl_ucp_context_t *ctx);
+
+static inline ucc_context_id_t
+ucc_tl_ucp_get_rank_key(ucc_tl_ucp_team_t *team, ucc_rank_t rank)
+{
+    size_t max_addrlen = team->addr_storage->max_addrlen;
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/tl/ucp/tl_ucp_ep.h:27]: :+1 fixed

File: src/components/tl/ucp/tl_ucp.h (hunk 4 above)
Comment (vspetrov) [reply src/components/tl/ucp/tl_ucp.h:76]: It was old flow (potentially eps could be stored on CTX) - removed that now. Maybe we will still implement exchange of addresses during ctx create (in global mode) but later.

File: src/components/tl/ucp/tl_ucp_context.c [hunk 6]
```diff
@@ -126,7 +129,12 @@ UCC_CLASS_INIT_FUNC(ucc_tl_ucp_context_t,
 
 UCC_CLASS_CLEANUP_FUNC(ucc_tl_ucp_context_t)
 {
+    ucc_status_t status;
     tl_info(self->super.super.lib, "finalizing tl context: %p", self);
+    while (UCC_OK != (status = ucc_tl_ucp_close_eps(self))) {
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/tl/ucp/tl_ucp_context.c:134]: 👍

File: src/components/tl/ucp/tl_ucp_ep.c (hunk 3 above)
Comment (vspetrov) [reply src/components/tl/ucp/tl_ucp_ep.c:71]: 👍

**Review by Sergei-Lebedev — APPROVED** [review]

**Pull request merged**